File: h3ui/video_assembly/track.py
"""Editable delivery order, separate from the immutable generation source chain.

Keys identify source clips or selected extension slots, never arbitrary files.
Legacy projects derive their original source/extension order without a write.
"""


def entries(project):
    runs = {r['id']: r for r in project['assembly']['runs']}
    rows = []
    for clip in project['assembly']['clips']:
        if clip.get('removed_at'):
            continue
        rows.append(('clip:' + clip['id'], clip, None))
        for extension in clip['extensions']:
            run = runs.get(extension.get('selected'))
            if (not extension.get('removed_at') and run and not run.get('removed_at')
                    and run['state'] == 'success' and run.get('extension') == extension['id']):
                rows.append(('extension:' + extension['id'], clip, run))
    return rows


def order(project):
    keys = [key for key, _, _ in entries(project)]
    stored = project['assembly'].get('track_order', [])
    # Removed/unselected slots disappear; newly available slots append once.
    return list(dict.fromkeys([key for key in stored if key in keys] + keys))


def validate(value, expected):
    if (not isinstance(value, list) or not all(isinstance(k, str) for k in value)
            or len(value) != len(set(value)) or set(value) != set(expected)):
        raise ValueError('视频轨道已变化或包含无效片段，请重新读取后排序；原草稿保留')
    return list(value)

File: static/studio/modes/video-assembly/track.js
/** Delivery slots; reordering never changes an extension's generation parent. */
export function trackItems(p){
  const rows=[];
  for(const c of p.assembly.clips.filter(c=>!c.removed_at)){
    rows.push({key:'clip:'+c.id,clip:c,id:c.id,name:c.name,url:c.url,cover:c.cover_url,seconds:c.end-c.start});
    c.extensions.forEach((e,i)=>{
      const run=p.assembly.runs.find(r=>r.id===e.selected&&r.extension===e.id&&r.state==='success'&&!r.removed_at);
      if(!e.removed_at&&run)rows.push({key:'extension:'+e.id,clip:c,extension:e,run,id:e.id,name:`${c.name} · 续接 ${i+1}`,url:run.url,cover:run.cover_url,seconds:run.report.duration});
    });
  }
  const byKey=new Map(rows.map(row=>[row.key,row]));
  return [...new Set([...(p.assembly.track_order||[]),...byKey.keys()])].filter(key=>byKey.has(key)).map(key=>byKey.get(key));
}

export function moveTrack(p,key,offset){
  const order=trackItems(p).map(row=>row.key),index=order.indexOf(key),to=index+offset;
  if(index<0||to<0||to>=order.length)return p;
  order.splice(to,0,...order.splice(index,1));
  return {...p,assembly:{...p.assembly,track_order:order}};
}
